Add half precision floating point support to `StreamPeer` and `FileAccess`

pafuent opened the pull request:
Closes godotengine/godot-proposals#5983

Adds put/get methods to `StreamPeer` that handles half precision floating point values.
Adds endode/decode half precision floating point to `marshalls`.

This PR is rebase into master of #78872. Also I added unit tests to it.

<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

pafuent committed:
Add half precision floating point support to `StreamPeer`

Closes godotengine/godot-proposals#5983

Adds put/get methods to `StreamPeer` that handles half precision
floating point values.
Adds endode/decode half precision floating point to `marshalls`.
Adds `get_half` and `store_half` to `FileAccess`

Co-Authored-By: "Alfonso J. Ramos" <theraot@gmail.com>

radiantgurl reviewed (commented):
Shouldnt we extend it to FileAccess and PackedByteArray now given the chance to do so?

pafuent commented:
`PackedByteArray` already have those:
`encode_half` binded [here](https://github.com/godotengine/godot/blob/master/core/variant/variant_call.cpp#L2431) and implemented [here](https://github.com/godotengine/godot/blob/master/core/variant/variant_call.cpp#L1002)
`decode_half` binded [here](https://github.com/godotengine/godot/blob/master/core/variant/variant_call.cpp#L2411) and implemented [here](https://github.com/godotengine/godot/blob/master/core/variant/variant_call.cpp#L840)

RadiantUwU commented:
Topic labels changed, we should change the title and the topics since now it also modifies FileAccess (core)

pafuent commented:
Friendly remainder

Calinou reviewed (approved):
Tested locally, it works as expected.

Thanks for adding unit tests :slightly_smiling_face: 

```gdscript
extends Node2D


func _ready() -> void:
	var fa := FileAccess.open("user://output.bin", FileAccess.WRITE_READ)
	fa.store_half(0.123456)
	fa.store_float(0.123456)
	fa.store_double(0.123456)

	fa.store_double(-12345)
	fa.store_float(-12345)
	fa.store_half(-12345)

	OS.shell_open(ProjectSettings.globalize_path("user://"))
	get_tree().quit()
```

Using [ImHex](https://imhex.werwolv.net/) to view the file written by Godot:

![image](https://github.com/user-attachments/assets/f450f125-a70a-4206-b0fc-acb9b64a9315)

![image](https://github.com/user-attachments/assets/2a9b962d-ff19-42dc-8bac-77bde9d95a9e)

Note the lack of precision on the second example, as half floats can't represent all integer values outside of `[-2048; 2048]`.

AThousandShips commented on tests/core/io/test_file_access.h:
```diff
@@ -107,6 +107,98 @@ TEST_CASE("[FileAccess] Get as UTF-8 String") {
+		const String file_path_new = TestUtils::get_data_path("floating_point_little_endian_new.bin");
+
+		Ref<FileAccess> f = FileAccess::open(file_path, FileAccess::READ);
+		REQUIRE(!f.is_null());
```
```suggestion
		REQUIRE(f.is_valid());
```

AThousandShips commented on tests/core/io/test_file_access.h:
```diff
@@ -107,6 +107,98 @@ TEST_CASE("[FileAccess] Get as UTF-8 String") {
+		CHECK_EQ(f->get_float(), value);
+
+		Ref<FileAccess> fw = FileAccess::open(file_path_new, FileAccess::WRITE);
+		REQUIRE(!fw.is_null());
```
```suggestion
		REQUIRE(fw.is_valid());
```

AThousandShips commented on tests/core/io/test_file_access.h:
```diff
@@ -107,6 +107,98 @@ TEST_CASE("[FileAccess] Get as UTF-8 String") {
+		const String file_path_new = TestUtils::get_data_path("floating_point_big_endian_new.bin");
+
+		Ref<FileAccess> f = FileAccess::open(file_path, FileAccess::READ);
+		REQUIRE(!f.is_null());
```
```suggestion
		REQUIRE(f.is_valid());
```

pafuent commented:
Friendly reminder

Faless reviewed (approved):
LGTM! Nice work! :rocket:

akien-mga commented:
It's now queued for merging, thanks for your patience :)

Repiteo commented:
Thanks!
